src/net.cpp

```diff
@@ -1028,6 +1067,15 @@ void ThreadSocketHandler2(void* parg)
                     // remove from vNodes
                     vNodes.erase(remove(vNodes.begin(), vNodes.end(), pnode), vNodes.end());
 
+                    // reduce score of the address used
+                    if (pnode->fInbound)
+                    {
+                        if (mapLocalHost[pnode->addrLocal].nScore > 1)
+                        {
+                           mapLocalHost[pnode->addrLocal].nScore--;
+                        }
```

Comment:
```suggestion
                        LOCK(cs_mapLocalHost);
                        auto iter = mapLocalHost.find(pnode->addrLocal);
                        
                        if (iter != mapLocalHost.end() && iter->second.nScore > 1)
                        {
                            --iter->second.nScore;
                        }
```

The value of `pnode->addrLocal` is untrusted input supplied by the peer. When we use the subscript operator (`[]`) on `std::map` objects, the map will create a new record if the key doesn't exist. This enables malicious nodes to potentially fill up `mapLocalHost` with bogus addresses. Instead, we can use `std::map::find()` to check for existence first. 

We also need to lock `cs_mapLocalHost` because other threads access and modify the map.

Reply:
Speaking of bogus entries, should we remove `pnode->addrLocal` from `mapLocalHost` when it's no longer valid? If the node's external IP address changes often, iteration over `mapLocalHost` in `GetLocal()` will gradually slow down as stale addresses build up. Does `nScore` ever drop below zero?

Reply:
I think we should remove stale address from the map.

Reply:
with the new map each peer has one vote, that is stored in the map, and on disconnect removed, alse each peer will only create one entry and not more.
I think i read something of connection limit of 250, or somewhere in that range, so the map will hold up to that ammount, i think that is acceptable now.

Reply:
Actually, we have addnodes that are supporting 750+ connections, but I think that is still ok.